src/abstract.js

```diff
@@ -0,0 +1,187 @@
+import { decorate } from './private/utils';
+
+
+export default function abstract(...args) {
+  let result;
+
+  if (args.length == 1) {
+    result = abstractclass(args[0]);
+  }
+  else {
+    result = decorate(handleDescriptor, args);
+  }
+
+  return result;
+}
+
+
+function handleDescriptor(target, attr, descriptor) {
+  let result;
+
+  const desc = {...descriptor};
+  if (typeof descriptor.value == 'function') {
+    result = abstractmethod(target, attr, desc);
+  } else {
+    throw new Error('@abstract can only be used on classes and methods');
+  }
+
+  return result;
+}
+
+
+function abstractmethod(target, attr, descriptor) {
+  const result = descriptor;
+
+  result.value = makeAbstract(attr);
+
+  return result;
+}
+
+
+function abstractclass(target) {
+  const result = new Function(
+    ['getSuper', 'classCallCheck', 'isConcrete', 'target'],
+    `
+    function ${target.name}() {
+      classCallCheck(this, ${target.name});
+
+      var isAbstract = this.constructor.name == '${target.name}';
+      var isSubclassConcrete = true;
+      if (!isAbstract) {
+        isSubclassConcrete = isConcrete(this.constructor);
+      }
+      if (isAbstract || !isSubclassConcrete) {
+        const parts = [
+          'abstract classes cannot be instantiated'
+        ];
+        if (!isSubclassConcrete) {
+          parts.push(
+            'did you forget to decorate ' + this.constructor.name + ' using @abstract?'
+          );
+        }
+
+        throw new TypeError(parts.join('\\n'));
+      }
+
+      getSuper(Object.getPrototypeOf(${target.name}.prototype), 'constructor', this).apply(this, arguments);
+    };
+
+    return ${target.name};
+    `
+  )(getSuper, classCallCheck, isConcrete, target);
+
+  inherit(result, target);
+
+  return result;
+}
+
+
+function isConcrete(constructor) {
+  let result = true;
+
+  // collect own properties along the inheritance chain
+  let map = {};
+  let ctor = constructor;
+  while (ctor.prototype) {
+    let objs = [ctor, ctor.prototype];
+
+    for (let index=0; index<objs.length; index++) {
+      for (const attr of Object.getOwnPropertyNames(objs[index])) {
+        if (!(attr in map)) {
+          const desc = Object.getOwnPropertyDescriptor(objs[index], attr); 
+          if (desc) {
+            map[attr] = desc;
+          }
+        }
+      }
+    }
+
+    ctor = Object.getPrototypeOf(ctor);
+  }
+
+  for (const attr in map) {
+    const desc = map[attr];
+
+    if (desc) {
+      const fun = desc.get || desc.set || typeof desc.value == 'function' ? desc.value : undefined;
+
+      if (fun && fun.__abstract) {
+        result = false;
+        break;
+      }
+    }
+  }
+
+  return result;
+}
+
+
+function makeAbstract(attr) {
+  const result = new Function(
+    `
+    return function ${attr}() {
+      throw new Error('derived classes must implement this');
```

Comment:
Can this error message be more descriptive?

Something like `${derivedClassName} does not implement abstract method ${methodName}()`